flink-table/flink-sql-client/src/main/java/org/apache/flink/table/client/cli/CliTableauResultView.java

```diff
@@ -122,14 +137,83 @@ private void checkAndCleanUpQuery(boolean cleanUpQuery) {
     private void printBatchResults(AtomicInteger receivedRowCount) {
         final List<RowData> resultRows = waitBatchResults();
         receivedRowCount.addAndGet(resultRows.size());
+
+        if (printBatchEmptySet(resultRows)) return;
+        printBatchTable(resultRows);
+        printBatchFooter(resultRows.size());
+    }
+
+    /**
+     * Print "Empty set" if the given <code>resultRows</code> is empty, otherwise will do nothing.
+     *
+     * @return false if <code>resultRows</code> is not empty
+     */
+    public boolean printBatchEmptySet(List<RowData> resultRows) {
+
+        if (!resultRows.isEmpty()) {
+            return false;
+        }
+
+        if (!resultDescriptor.isPrintQueryTimeCost()
+                || DEFAULT_QUERY_BEGIN_TIME == queryBeginTime) {
+            terminal.writer().println("Empty set");
+        } else {
+            String timeCost =
+                    calculateTimeCostInPrintFormat(queryBeginTime, System.currentTimeMillis());
+            terminal.writer().println("Empty set" + timeCost);
+        }
+        terminal.writer().flush();
+        return true;
+    }
+
+    /**
+     * Print table with column names and borders
+     *
+     * @return the row number printed in the table
+     */
+    public void printBatchTable(List<RowData> resultRows) {
```

Comment:
public -> private

printBatchTable -> printBatchResultsInternal

BTW, I think we can merge printBatchTable and printBatchFooter into one. WDYT?

Reply:
using two methods is more about code readability. Table and footer are two different concepts in this case. I would suggest keep them isolated and also for better maintainability, e.g. footer might be changed without need to know anything about the table. WDYT?